pkg/util/context/plancache.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package context
+
+import (
+	"sync"
+
+	"github.com/pingcap/errors"
+)
+
+// PlanCacheType is the flag of plan cache
+type PlanCacheType int
+
+const (
+	// DefaultNoCache no cache
+	DefaultNoCache PlanCacheType = iota
+	// SessionPrepared session prepared plan cache
+	SessionPrepared
+	// SessionNonPrepared session non-prepared plan cache
+	SessionNonPrepared
+)
+
+// PlanCacheTracker PlanCacheTrackerf `PlanCacheTracker`.
+type PlanCacheTracker struct {
+	UseCache            bool
```

Comment:
Could we make `PlanCacheTracker` a thread-safe? I schedule a PR #52727 to make it to avoid some data race if `BuildContext` is used in concurrency.

If we can do it in this PR, I can close #52727

Reply:
Theoretically, even `BuildContext` may be used concurrently, it only reads the `UseCache` right? Then the `BuildContext` is still thread-safe even if we change nothing :thinking: .

I agree that it's always better to make it thread-safe. Making change here can also make it safe even if the `PlanContext`, `BuildContext` is used concurrently :+1: .

Let me do it.